### PR TITLE
new URL parsing

### LINKED ISSUE #541: UrlStr gives unhelpful validation errors
# Bug

For bugs/questions:
* OS: macOS
* Python version `import sys; print(sys.version)`: 3.7.1
* Pydantic version `import pydantic; print(pydantic.VERSION)`: StrictVersion ('0.25')

Here's a short example:

```py
>>> from pydantic import BaseModel, urlstr
>>> class DatabaseConfig(BaseModel):
...     url: urlstr(schemes={'postgresql'}, require_tld=False) = ...
... 
>>> DatabaseConfig(url='postgresql://diary_test@localhost/diary_test')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 235, in __init__
    object.__setattr__(self, '__values__', self._process_values(data))
  File "pydantic/main.py", line 438, in _process_values
    return validate_model(self, input_data)  # type: ignore
  File "pydantic/main.py", line 635, in validate_model
    raise ValidationError(errors)
pydantic.error_wrappers.ValidationError: 1 validation error
url
  url string does not match regex (type=value_error.url.regex)
```

So, couple of problems:

- This url is valid, it works ;-) 
- "does not match regex" isn't at all helpful to the end user, we should at least include the pattern being matched.

### LINKED ISSUE #603: UrlStr does not match e.g. kubernetes URLs
This URL does not pass validation: `http://action-server/webhook`. Any reason it should not?

Related to #246

### PR DESCRIPTION
**Pull request by samuelcolvin**
# Change Summary

Remove `DSN` and move to `UrlStr` to `AnyUrl` , and much more.

## Related issue number

fix #603, fix #541

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.rst` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by samuelcolvin** [comment]
For me this is ready except for lots of docs to write.

Feedback very welcome, preferably before I write all the docs.

**Comment by samuelcolvin** [comment]
### public classes and functions

* `AnyUrl`: any scheme allowed, userinfo is optional, tld not required, max length 2^16
* `AnyHttpUrl`: schema `http` or `https`, userinfo is optional, tld not required, max length 2^16
* `HttpUrl`: schema `http` or `https`, userinfo is optional, tld required, max length 2083
* `PostgresDsn`: schema `postgres` or `postgresql`, userinfo required, tld not required, max length 2^16
* `RedisDsn`: schema `postgres`, userinfo required, tld not required, max length 2^16
* `urlstr(strip_whitespace: bool = True, min_length: int = 1, max_length: int = 2 ** 16, tld_required: bool = True, allowed_schemes: Optional[Set[str]] = None)` maybe should rename to `stricturl()`?

### Other changes

* international domains are allowed but encoded with punycode: `'https://www.аррӏе.com/' -> 'https://www.xn--80ak6aa92e.com/'`, `'https://exampl£e.org' -> 'https://xn--example-gia.org'`
* `AnyUrl` and subclasses are a subclass of `str`, but have extra properties: `'scheme', 'user', 'password', 'host', 'tld', 'host_type', 'port', 'path', 'query', 'fragment'` to allow easier interpretation or further validation
* error messages for invalid urls should be more helpful, eg. saying which part is invalid

### known edge cases

underscores are now allowed in all parts of a domain except the tld. Technically I think this might be slightly wrong - I think in theory the hostname cannot have underscores but subdomains can. However, consider the following two cases:

* `exam_ple.co.uk` hostname is `exam_ple`, should not be allowed as there's an underscore in there
* `foo_bar.example.com` hostname is `example` should be allowed since the underscore is in the subdomain

Without having an exhaustive list of TLDs it would be impossible to differentiate between these two. Therefore underscores are allowed, you could do further validation in a validator if you wanted.

Also, chrome currently accepts `http://exam_ple.com` as a URL, so we're in good (or at least big) company.

File: tests/test_networks.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    'value,err_type,err_msg,err_ctx',
+    [
+        (
+            'ftp://example.com/',
+            'value_error.url.scheme',
+            'URL scheme not permitted',
+            {'allowed_schemes': {'https', 'http'}},
+        ),
+        ('http://foobar/', 'value_error.url.host', 'URL host invalid, top level domain required', None),
+        ('http://localhost/', 'value_error.url.host', 'URL host invalid, top level domain required', None),
```
Comment (cjw296) [inline tests/test_networks.py]: ...but these work in browsers, which is problematic.

Comment (samuelcolvin) [reply tests/test_networks.py]: Then use `AnyHttpUrl`.

File: tests/test_networks.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        '\u001f@example.com',
+        '"@example.com',
+        '\"@example.com',
+        ',@example.com',
+        'foobar <foobar<@example.com>',
+    ],
+)
+def test_address_invalid(value):
+    with pytest.raises(ValueError):
+        validate_email(value)
```
Comment (cjw296) [inline tests/test_networks.py]: What are the error messages like on these? I wonder how many of them actually end up being valid as far as the RFCs in play are concerned? *sigh*

Comment (samuelcolvin) [reply tests/test_networks.py]: This hasn't changed, just moved the tests. The error message will be given the [email-validator](https://pypi.org/project/email-validator/) package.

**Pull request merged**